controllers/legacy_resource_quota_controller.go

```diff
@@ -0,0 +1,141 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// LegacyResourceQuotaReconciler reconciles namespaces and synchronizes their resource quotas
+type LegacyResourceQuotaReconciler struct {
+	client.Client
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+
+	OrganizationLabel string
+
+	ResourceQuotaAnnotationBase string
+	DefaultResourceQuotas       map[string]corev1.ResourceQuotaSpec
+
+	LimitRangeName    string
+	DefaultLimitRange corev1.LimitRangeSpec
+}
+
+func (r *LegacyResourceQuotaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+	l.Info("Reconciling Namespace")
+
+	var ns corev1.Namespace
+	if err := r.Get(ctx, req.NamespacedName, &ns); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	if ns.DeletionTimestamp != nil {
+		l.Info("Namespace is being deleted, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	if _, ok := ns.Labels[r.OrganizationLabel]; !ok {
+		l.Info("Namespace does not have organization label, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	var errs []error
+	for name, s := range r.DefaultResourceQuotas {
+		spec := *s.DeepCopy()
+
+		var storageQuotas corev1.ResourceList
+		if sqa := ns.Annotations[fmt.Sprintf("%s/%s.storageclasses", r.ResourceQuotaAnnotationBase, name)]; sqa != "" {
+			err := json.Unmarshal([]byte(ns.Annotations[fmt.Sprintf("%s/%s.storageclasses", r.ResourceQuotaAnnotationBase, name)]), &storageQuotas)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("failed to unmarshal storage classes: %w", err))
+				storageQuotas = make(corev1.ResourceList)
+			}
+		} else {
+			storageQuotas = make(corev1.ResourceList)
+		}
+
+		rq := &corev1.ResourceQuota{
+			ObjectMeta: ctrl.ObjectMeta{
+				Name:      name,
+				Namespace: ns.Name,
+			},
+		}
+		op, err := controllerutil.CreateOrUpdate(ctx, r.Client, rq, func() error {
+			for k := range spec.Hard {
+				an := fmt.Sprintf("%s/%s.%s", r.ResourceQuotaAnnotationBase, name, strings.ReplaceAll(string(k), "/", "_"))
+				if strings.Contains(string(k), "storageclass.storage.k8s.io") {
+					if _, ok := storageQuotas[k]; ok {
+						spec.Hard[k] = storageQuotas[k]
+						continue
+					}
+				}
+
+				if a := ns.Annotations[an]; a != "" {
```

Comment:
Should we skip this for storage quotas to ensure the behavior is unchanged compared to the Kyverno policy?

Currently the following annotation would have different effects between the Kyverno policy and this policy (ignoring that it might be too long for a valid annotation key):

```
resourcequota.appuio.io/rbd-storagepool-cluster.storageclass.storage.k8s.io_requests.storage: '50Gi'
```

Reply:
Good catch! I'll skip it.

Reply:
I put the conitnue in the wrong place but i think this is one of the few cases where `else if` makes the control flow clearer.

Reply:
In this case it's certainly easier to reason that the implementation is correct with the else-if, I looked at the `continue` for a bit during the review, but the new version is much easier to read.